prow/cmd/crier/main.go

```diff
@@ -58,7 +58,7 @@ type options struct {
 }
 
 func (o *options) Validate() error {
-	if o.gerritWorkers > 0 {
+	if o.gerritWorkers > 1 {
 		// TODO(krzyzacy): try to see how to handle racy better for gerrit aggregate report.
```

Comment:
didn't we decide to just have a single "actually send the report to gerrit" goroutine and a channel?

Reply:
how can you tell if it's a gerrit job :-] and we don't want to do that for other reporters :-]

Reply:
quick terrible mock:
```go
func ParallelGenericReportHandler(pj *ProwJob) {
    for _, reporter := range reporters {
        go reporter.Report(pj)
    }
}

func (gr *GerritReporter) Report(pj *ProwJob) {
      gr.reportChan <- pj
}

// should be started somewhere in general initialization
func (gr *GerritReporter) ReportLoop() {
    for {
        select {
        case <- gr.quit:
            return
        case pj := <-gr.reportChan:
           gr.actuallyReport(pj)
    }
}
```

Reply:
then the reporter is doing the same work as crier controller :-]

Ideally I was also thinking of introduce another CRD or struct like batch which can hold a list of prowjobs, which will make the logic more clear I think?

Reply:
the racy part is not only report.. `ShouldReport`, and also update states to apiserver, are also within the critical section :-]

/this-is-fine

Reply:
er, what I mean is the controller can generally be parallel, but the gerrit reporter implementation should merge and serially handle these states, which can easily be not racy.

> the racy part is not only report.. ShouldReport, and also update states to apiserver, are also within the critical section :-]

yes exactly, put that in the actuallyReport() call, bad naming aside

Reply:
yeah I understand your idea, I don't want to put the crier logic (apiserver calls) into gerrit reporter though? (reporter doesn't handle state changes, crier does)

Also we probably want reporting for different patchset to happen in parallel.. Maybe a per-patchset lock is sufficient.

Reply:
> (reporter doesn't handle state changes, crier does)

that seems like a strange distinction. Reporting is handling a state change ...? it's handling the change from not needing reporting, to needing reporting :^) 

The gerrit reporter just needs to batch and merge those.

> Also we probably want reporting for different patchset to happen in parallel.. Maybe a per-patchset lock is sufficient.

 We're not handling high enough QPS to worry about this yet. We probably [gerrit] report a few times an hour, not 1000s of time a second.

Reply:
yeah... so... let's worry about this later :-)